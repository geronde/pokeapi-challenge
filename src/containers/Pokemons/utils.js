import { get } from 'lodash';

export const paginate = (
  totalItems,
  currentPage = 1,
  pageSize = 20,
) => {
  const totalPages = Math.ceil(totalItems / pageSize);
  let cp = 1;

  if (currentPage < 1) {
    cp = 1;
  } else if (currentPage > totalPages) {
    cp = totalPages;
  }

  const startPage = 1;
  const endPage = totalPages;

  const startIndex = (cp - 1) * pageSize;
  const endIndex = Math.min(
    startIndex + pageSize - 1,
    totalItems - 1,
  );

  return {
    totalItems,
    currentPage: cp,
    pageSize,
    totalOfPages: totalPages,
    startPage,
    endPage,
    startIndex,
    endIndex,
  };
};

export const getIdFromUrl = (url) => url.split('/').slice(-2)[0];

export const formatEvolutions = (pokemon) => {
  const evoChain = [];
  let evoData = pokemon.chain;

  do {
    const numberOfEvolutions = evoData.evolves_to.length;

    evoChain.push({
      species_name: get(evoData, 'species.name'),
    });

    if (numberOfEvolutions > 1) {
      for (let index = 1; index < numberOfEvolutions; index += 1) {
        evoChain.push({
          species_name: get(
            evoData.evolves_to[Number.parseInt(index, 10)],
            'species.name',
          ),
        });
      }
    }

    evoData = { ...evoData.evolves_to[0] };
  } while (evoData !== undefined && 'evolves_to' in evoData);

  return evoChain;
};
