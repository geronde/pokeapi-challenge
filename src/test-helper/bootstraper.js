import faker from 'faker';

export const generateTestPokemons = (n = 15, isType) => {
  const pokemons = [];

  for (let i = 0; i < n; i++) {
    let poke = { name: faker.random.word(), url: `fake-url/${i}/` };
    if (isType) {
      poke = { pokemon: { ...poke } };
    }
    pokemons.push(poke);
  }
  return pokemons;
};
