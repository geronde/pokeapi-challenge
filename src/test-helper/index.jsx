import { render as rtlRender } from '@testing-library/react';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';

export const render = (ui, { ...renderOptions } = {}) =>{
    function Wrapper({ children }) {
      return <BrowserRouter>{children}</BrowserRouter>
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
  }