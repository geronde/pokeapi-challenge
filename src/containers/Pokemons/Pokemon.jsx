import React, { useState, useEffect, useContext } from "react";
import styled from "styled-components";
import { uniqueId, isEmpty, get } from "lodash";
import { Link } from "react-router-dom";

import AppLoader from "../../components/Loader";
import Header from "../../components/Header";
import Footer from "../../components/Footer";
import { getIdFromUrl } from "./utils";

import { LocaleContext } from '../LocaleProvider/index'
import messages from './messages';


const Wrapper = styled.div`
  width: 100%;
  min-height: 50vh;
  position: relative;
  display: grid;
  grid-template-rows: 1fr 6fr 1fr;
  grid-template-areas: "header header header" "content content content" "footer footer footer";
  background: #e2e1e0;
  main {
    grid-area: content;
    display: flex;
    flex-direction: column;
    width: 55%;
    margin: 30px auto;
    color: #000;
  }
  .pokemon-detail {
    padding: 20px;
    width: 80%;
    margin: 0 auto;
    background-color: #fff;
    display: flex;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  }
  img {
    width: 150px;
  }
  .detail {
    padding: 20px;
    background-color: #eee;
    width: 70%;
    display: column;
  }
  .name {
    font-size: 18px;
    font-family: Roboto;
    font-weight: bold;
    text-transform: uppercase;
  }
  .type,
  .abilities,
  .evolution {
    display: flex;
    flex-direction: column;
    justify-content: center;
  }
  .type div,
  .abilities div,
  .evolution div {
    margin-left: -35px;
  }
`;


const Pokemon = (props) => {
  const [pokemon, setPokemon] = useState({});

  useEffect(() => {
    const fetchPokemon = async (url, evolutionUrl) => {
      const res = await fetch(url);
      const evol = await fetch(evolutionUrl);
      const evolutionResponse = await evol.json();
      const response = await res.json();

      setPokemon({ ...response, ...evolutionResponse });
    };

    const id = props.match.params.id;
    fetchPokemon(
      `https://pokeapi.co/api/v2/pokemon/${id}`,
      `https://pokeapi.co/api/v2/evolution-chain/${id}/`
    );
  }, [props.match.params.id]);

  const formatttedEvoltions = () => {
    let evoChain = [];
    let evoData = pokemon.chain;

    do {
      let numberOfEvolutions = evoData.evolves_to.length;

      evoChain.push({
        species_name: get(evoData, "species.name"),
      });

      if (numberOfEvolutions > 1) {
        for (let i = 1; i < numberOfEvolutions; i++) {
          evoChain.push({
            species_name: get(evoData.evolves_to[i], "species.name"),
          });
        }
      }

      evoData = evoData.evolves_to[0];
    } while (evoData !== undefined && evoData.hasOwnProperty("evolves_to"));

    return evoChain;
  };

  const {translate} = useContext(LocaleContext);


  if (isEmpty(pokemon)) return <AppLoader/>

  return (
    <Wrapper>
      <Header>{pokemon.name}</Header>
      <main>
        <section className="pokemon-detail">
          <article className="img">
            <img alt={pokemon.name} src={pokemon.sprites.front_default} />
          </article>
          <article className="detail">
            <div className="name">{pokemon.name}</div>
            <ul className="type">
  <div>{translate(messages.types)}</div>
              {pokemon.types.map((type) => (
                <Link to={`/types/${getIdFromUrl(type.type.url)}`}><li key={uniqueId(`${pokemon.name}`)}>{type.type.name}</li></Link>
              ))}
            </ul>
            <ul className="abilities">
              <div>{translate(messages.abilities)}</div>
              {pokemon.abilities.map((ab) => (
                <li key={ab.slot}>{ab.ability.name}</li>
              ))}
            </ul>
            <ul className="evolution">
              <div>{translate(messages.evolutions)}</div>
              {formatttedEvoltions().map((evo) => (
                <li key={uniqueId()}>{evo.species_name}</li>
              ))}
            </ul>
          </article>
        </section>
      </main>
      <Footer/>
    </Wrapper>
  );
};

export default Pokemon;
