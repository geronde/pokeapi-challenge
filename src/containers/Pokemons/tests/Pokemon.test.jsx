import React from 'react';
import { act, screen, cleanup } from '@testing-library/react'

import { render } from '../../../test-helper';

import Pokemon from '../Pokemon';
import * as utils from '../utils'

afterEach(cleanup);

describe('Pokemon', () => {
    let pokemon
    beforeEach(() => {
        pokemon = jest.fn(() => Promise.resolve({ name: 'pokemon-name', sprites: {front_default: 'fake-img.png'}, types: [{ type: {name: 'type-1', url: 'fake-url/3/'}}, { type: {name: 'type-2', url: 'fake-url/4/'}}], abilities: [{slot: '1', ability: {name: 'abilit-1', url: 'fake-url/8'}}, {slot: '15',ability: {name: 'abilit-2', url: 'fake-url/78'}}]}));

        jest.spyOn(utils, 'formatEvolutions').mockReturnValue([{species_name: 'Specie-1'}, {species_name: 'Specie-2'},{species_name: 'Specie-3'}])
    });

    afterEach(() => {
        jest.resetAllMocks();
        jest.resetAllMocks();
    });

it('should render a loader if pokemon is empty', async () => {
    jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: jest.fn(() => Promise.resolve([])) });
    await act(async () => {
        const type = {params: {type: '2'}}
        const { asFragment } = render(<Pokemon match={type} />)
        expect(asFragment()).toMatchSnapshot();
    })
    });

    it('should call fetch on page load and display pokemon infromation', async () => {
        jest.spyOn(window, 'fetch').mockReturnValueOnce({ json: pokemon }).mockReturnValue({json: jest.fn(()=> Promise.resolve(['evolutions']))});
        await act(async () => {
            const id = {params: {id: '4'}}
            render(<Pokemon match={id} />)
        })  
        expect(window.fetch).toHaveBeenCalledTimes(2)
        expect(screen.getAllByText('pokemon-name')[0]).toBeInTheDocument()
        expect(utils.formatEvolutions).toHaveBeenCalledTimes(1)
        expect(screen.getAllByRole('article')[1]).toMatchSnapshot();
    });
})