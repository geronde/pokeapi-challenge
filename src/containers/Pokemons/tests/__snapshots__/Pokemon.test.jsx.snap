// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon should call fetch on page load and display pokemon infromation 1`] = `
<article
  class="detail"
>
  <div
    class="name"
  >
    pokemon-name
  </div>
  <ul
    class="type"
  >
    <div>
      Types
    </div>
    <a
      href="/types/3"
    >
      <li>
        type-1
      </li>
    </a>
    <a
      href="/types/4"
    >
      <li>
        type-2
      </li>
    </a>
  </ul>
  <ul
    class="abilities"
  >
    <div>
      Abilities
    </div>
    <li>
      abilit-1
    </li>
    <li>
      abilit-2
    </li>
  </ul>
  <ul
    class="evolution"
  >
    <div>
      Evolutions
    </div>
    <li>
      Specie-1
    </li>
    <li>
      Specie-2
    </li>
    <li>
      Specie-3
    </li>
  </ul>
</article>
`;

exports[`Pokemon should render a loader if pokemon is empty 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
  >
    <div
      aria-busy="true"
      class=""
    >
      <svg
        aria-label="audio-loading"
        height="100"
        stroke="#0E1111"
        viewBox="0 0 44 44"
        width="100"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
          stroke-width="2"
        >
          <circle
            cx="22"
            cy="22"
            r="1"
          >
            <animate
              attributeName="r"
              begin="0s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.165, 0.84, 0.44, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 20"
            />
            <animate
              attributeName="strokeOpacity"
              begin="0s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.3, 0.61, 0.355, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 0"
            />
          </circle>
          <circle
            cx="22"
            cy="22"
            r="1"
          >
            <animate
              attributeName="r"
              begin="-0.9s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.165, 0.84, 0.44, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 20"
            />
            <animate
              attributeName="strokeOpacity"
              begin="-0.9s"
              calcMode="spline"
              dur="1.8s"
              keySplines="0.3, 0.61, 0.355, 1"
              keyTimes="0; 1"
              repeatCount="indefinite"
              values="1; 0"
            />
          </circle>
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
